### ISSUE TITLE
BUG: zod zccepts invalid url: `z.string().url().safeParse("a:")`

### ISSUE THREAD

**Issue by valerii15298** [colinhacks/zod#2477]
Current behaviour:
```
// zod 3.21.4
console.log(z.string().url().safeParse("a:")); // {success: true}
```
I expect validation to fail there

**Comment by JaneJeon** [comment]
It's because the URL is being validated with just a `new URL()`: https://github.com/colinhacks/zod/blob/master/src/types.ts#L757

Probably should add checks to make sure only certain protocols (http/https?) get through, and that it has a valid, non-empty host

**Comment by valerii15298** [comment]
@JaneJeon yeah...  does not seem right to me using native URL class... I would choose validator.js for url validation really. Because with URL class we do not have any flexibility... For example class-validator uses validator.js url validation too.
I guess this is related: https://github.com/colinhacks/zod/issues/120#issuecomment-675674977
@colinhacks
is there a place for negotiating url validation or what would be workarounds for that if I wanna validate http urls? Should I use `refine` for it or maybe there is a better solution?

**Comment by JaneJeon** [comment]
Still unaddressed

**Comment by JacobWeisenburger** [comment]
Is this what you are looking for?

```ts
const schema = z.string().startsWith( 'http' ).url()
console.log( schema.safeParse( 'a:' ).success ) // false
console.log( schema.safeParse( 'http:a' ).success ) // true
```

If you found my answer satisfactory, please consider supporting me. Even a small amount is greatly appreciated. Thanks friend! 🙏
https://github.com/sponsors/JacobWeisenburger
